File: saksflyt/src/main/java/no/nav/melosys/saksflyt/ProsessinstansBehandlerDelegate.java
package no.nav.melosys.saksflyt;

import java.time.LocalDateTime;
import java.util.Collection;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

import no.nav.melosys.saksflytapi.domain.ProsessStatus;
import no.nav.melosys.saksflytapi.domain.Prosessinstans;
import no.nav.melosys.saksflytapi.domain.SedLåsReferanse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

@Component
public class ProsessinstansBehandlerDelegate {

    private static final Logger log = LoggerFactory.getLogger(ProsessinstansBehandlerDelegate.class);

    private final ProsessinstansBehandler prosessinstansBehandler;
    private final ProsessinstansRepository prosessinstansRepository;

    public ProsessinstansBehandlerDelegate(ProsessinstansBehandler prosessinstansBehandler, ProsessinstansRepository prosessinstansRepository) {
        this.prosessinstansBehandler = prosessinstansBehandler;
        this.prosessinstansRepository = prosessinstansRepository;
    }

    public void behandleProsessinstans(Prosessinstans prosessinstans) {
        oppdaterStatusOmSkalPåVent(prosessinstans);
        if (!prosessinstans.erPåVent()) {
            prosessinstansBehandler.behandleProsessinstans(prosessinstans);
        }
    }

    void oppdaterStatusOmSkalPåVent(Prosessinstans prosessinstans) {
        if (skalSettesPåVent(prosessinstans)) {
            prosessinstans.setStatus(ProsessStatus.PÅ_VENT);
            prosessinstans.setEndretDato(LocalDateTime.now());
            prosessinstansRepository.save(prosessinstans);
            log.info("Prosessinstans {} satt på vent", prosessinstans.getId());
        }
    }

    /*
    Settes på vent om det finnes en prosessinstans med samme referanse,
     men ikke lik identifikator i prosess (ikke på vent/ferdig).

    Settes ikke på vent om
        1. Prosessinstansen ikke har en låsreferanse
        2. Det finnes ingen prosessinstans med samme referanse
        3. Det finnes en prosessinstans med lik referanse og identifikator.
     */
    private boolean skalSettesPåVent(Prosessinstans prosessinstans) {
        if (prosessinstans.getLåsReferanse() == null) {
            return false;
        }

        final var låsReferanse = new SedLåsReferanse(prosessinstans.getLåsReferanse());
        final var aktiveLåsReferanser = finnAndreAktiveLåsMedSammeReferanse(prosessinstans.getId(), låsReferanse);

        if (aktiveLåsReferanser.contains(låsReferanse)) {
            return false;
        }
        return aktiveLåsReferanser.stream().anyMatch(
            sedLåsReferanse -> sedLåsReferanse.getReferanse().equals(låsReferanse.getReferanse()));
    }

    private Collection<SedLåsReferanse> finnAndreAktiveLåsMedSammeReferanse(UUID id, SedLåsReferanse låsReferanse) {
        return prosessinstansRepository.findAllByIdNotAndStatusNotInAndLåsReferanseStartingWith(
            id, Set.of(ProsessStatus.PÅ_VENT, ProsessStatus.FERDIG), låsReferanse.getReferanse()
        )
            .stream()
            .map(p -> new SedLåsReferanse(p.getLåsReferanse()))
            .collect(Collectors.toSet());
    }
}

File: saksflyt/src/main/kotlin/no/nav/melosys/saksflyt/ProsessinstansBehandlerDelegate.kt
package no.nav.melosys.saksflyt

import mu.KotlinLogging
import no.nav.melosys.saksflytapi.domain.LåsReferanse
import no.nav.melosys.saksflytapi.domain.LåsReferanseFactory
import no.nav.melosys.saksflytapi.domain.ProsessStatus
import no.nav.melosys.saksflytapi.domain.Prosessinstans
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.util.*

private val log = KotlinLogging.logger { }

@Component
class ProsessinstansBehandlerDelegate(
    private val prosessinstansBehandler: ProsessinstansBehandler,
    private val prosessinstansRepository: ProsessinstansRepository
) {

    fun behandleProsessinstans(prosessinstans: Prosessinstans) {
        oppdaterStatusOmSkalPåVent(prosessinstans)
        if (!prosessinstans.erPåVent()) {
            prosessinstansBehandler.behandleProsessinstans(prosessinstans)
        }
    }

    fun oppdaterStatusOmSkalPåVent(prosessinstans: Prosessinstans) {
        if (skalSettesPåVent(prosessinstans)) {
            prosessinstans.status = ProsessStatus.PÅ_VENT
            prosessinstans.endretDato = LocalDateTime.now()
            prosessinstansRepository.save(prosessinstans)
            log.info("Prosessinstans {} med låsreferanse {} satt på vent", prosessinstans.id, prosessinstans.låsReferanse)
        }
    }

    /*
    Settes på vent om det finnes en prosessinstans med samme referanse, som ikke er på vent/ferdig.

    Settes ikke på vent om
        1. Prosessinstansen ikke har en låsreferanse
        2. Det finnes ingen aktiv prosessinstans med samme gruppe prefiks
     */
    private fun skalSettesPåVent(prosessinstans: Prosessinstans): Boolean {
        if (prosessinstans.låsReferanse == null) {
            return false
        }

        val låsReferanse: LåsReferanse = LåsReferanseFactory.lagLåsReferanse(prosessinstans.hentLåsReferanse)
        val andreAktiveLåsMedSammeGruppePrefiks = finnAndreAktiveLåsMedSammegruppePrefiks(prosessinstans.id!!, låsReferanse.gruppePrefiks)
        log.info { "Låsreferanse: ${prosessinstans.låsReferanse} Andre aktive med samme gruppe prefiks: $andreAktiveLåsMedSammeGruppePrefiks" }
        return låsReferanse.skalSettesPåVent(andreAktiveLåsMedSammeGruppePrefiks)
    }

    internal fun finnAndreAktiveLåsMedSammegruppePrefiks(id: UUID, låsReferanseStarterMed: String): Collection<String> =
        prosessinstansRepository.findAllByIdNotAndStatusNotInAndLåsReferanseStartingWith(
            id, setOf(ProsessStatus.PÅ_VENT, ProsessStatus.FERDIG), låsReferanseStarterMed
        ).map { it.låsReferanse }
}
